### PR TITLE
Implemented functional notifications bar

### PR DESCRIPTION
**Pull request by SR95**
Implemented socket.io, and integrated it with cors to update notifications for product sales.

### CODE REVIEW

File: server/bin/www [hunk 1]
```diff
@@ -33,17 +37,47 @@ var server = http.createServer(app);
...
   });
   server.on("error", onError);
   server.on("listening", onListening);
 });
 
+/**
+ * Socket setup
+ */
+const io = socketio(server);
+
```
Comment (aeciorc) [inline server/bin/www:56]:
```suggestion
const socketClients = {} // keep a list of clients (map userId -> socketId)
```

File: server/bin/www [hunk 2]
```diff
@@ -33,17 +37,47 @@ var server = http.createServer(app);
...
+io.on("connection", socket => {
+  console.log("We have a new connection.");
+  let encodedCookie = socket.handshake.headers.cookie;
+  var parsedCookies = cookie.parse(encodedCookie);
+
+  // User data of current Socket User stored in userData
+  // User data includes: userName, userEmail and userId
+  var userData = authJWTCheck(parsedCookies);
+
```
Comment (aeciorc) [inline server/bin/www:74]: if the JWT is invalid, drop the connection ( emmit "disconnect")
else, add to the clients object: `socketClients[userData.userId] = socket.id`

File: server/bin/www [hunk 3]
```diff
@@ -33,17 +37,47 @@ var server = http.createServer(app);
...
+  socket.on("getNotifications", async (message, callback) => {
+    // userNotifications gets all the notifications for a certain user
+    // userNotifications.dataValues = all the data values
+    // dataValues are: productId, userId, previousPrice, currentPrice, dismissed
+    let userNotifications = await notificationHandler.getProductsForNotifications(
+      userData.userId
+    );
+
+    callback(userNotifications);
+  });
```
Comment (aeciorc) [inline server/bin/www]:
```suggestion

```

File: server/bin/www [hunk 4]
```diff
@@ -33,17 +37,47 @@ var server = http.createServer(app);
...
+
+  socket.on("disconnect", () => {
+    console.log("User has left.");
+  });
```
Comment (aeciorc) [inline server/bin/www]: here you would remove the socket from the client object

File: server/bin/www [hunk 5]
```diff
@@ -33,17 +37,47 @@ var server = http.createServer(app);
...
+});
+
```
Comment (aeciorc) [inline server/bin/www:79]: For notifying clients, you want to add a method that takes the notifications and a userId. It should be used by the cronJob E.g:
```js
const sendNotification = (notifications, userId) =>{
const socketId = socketClients[userId]
if (!socketId) return 
 io.to(`${socketId}`).emit("notificationsEvent", notifications);

}
```

Then on the **client** side, the socket will listen to that particular event. E.g

```js
 socket.on("notificationsEvent", notifications => {
    //update state from the notifications
    });
```

**Review by aeciorc — CHANGES_REQUESTED** [review]

File: server/bin/www (hunk 4 above)
Comment (SR95) [reply server/bin/www]: https://github.com/hatchways/team-penne/pull/57#discussion_r414007157

By client object, you mean the object socketClients you suggested in this comment right?

**Pull request merged**